src/Microsoft.Extensions.DependencyInjection.Abstractions/IDesignTimeServiceProviderFactory.cs

```diff
@@ -0,0 +1,22 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// A factory for creating the application's <see cref="IServiceProvider"/> at design time. Implement this interface
+    /// to enable design-time services that need access to the application's services. Design-time services will
+    /// automatically discover implementations of this interface.
+    /// </summary>
+    public interface IDesignTimeServiceProviderFactory
```

Comment:
It is a factory with a specific purpose. Since tooling will be looking for implementations of this interface for that specific purpose it would be really problematic if arbitrary implementations used the same interface. 

We could create an interface that is just a general purpose factory and have this one derive from it, but not sure what the scenario is.